lib/AST/ArchetypeBuilder.cpp

```diff
@@ -1030,6 +1036,37 @@ auto ArchetypeBuilder::resolveArchetype(Type type, PotentialArchetype *basePA)
   return nullptr;
 }
 
+auto ArchetypeBuilder::FIXMEresolveArchetypeOrTypeThroughTypeAlias(
+    Type type, PotentialArchetype *basePA)
+    -> llvm::PointerUnion<PotentialArchetype *, Type> {
+  auto pa = resolveArchetype(type, basePA);
+  if (!pa)
+    return type;
+
+  pa = pa->getRepresentative();
+  if (!pa->getParent() || !pa->getTypeAliasDecl())
```

Comment:
Hmm... it occurs to me that there might be direct callers to `addSameTypeRequirementBetweenArchetypes` that unknowingly end up with a `typealias` for one of the archetypes. Should this logic be sunk down into that function instead?